backend/gsr_booking/api_wrapper.py

```diff
@@ -532,38 +311,175 @@ def get_affiliation(self, email):
-
-        lc_end = lc_start + datetime.timedelta(days=1)
-
-        # filters for all reservations for the given date
-        reservations = GSRBooking.objects.filter(
-            gsr__in=GSR.objects.filter(kind=GSR.KIND_LIBCAL),
-            start__gte=lc_start,
-            end__lte=lc_end,
-            is_cancelled=False,
-            user=user,
-        )
-        total_minutes = 0
-        for reservation in reservations:
-            # accumulates total minutes over all reservations
-            total_minutes += int((reservation.end.timestamp() - reservation.start.timestamp()) / 60)
-        # 120 minutes per day
-        return {CreditType.LIBCAL.value: 120 - total_minutes}
+            reservation = gsr_booking.reservation
+            if all(booking.is_cancelled for booking in reservation.gsrbooking_set.all()):
+                reservation.is_cancelled = True
+                reservation.save()                
+        else:
+            try: 
+                self.WBW.cancel_room(booking_id, user)
+            except APIError:
+                try: 
+                    self.LBW.cancel_room(booking_id, user)
+                except APIError:
+                    raise APIError("Error: Unknown booking id")
+                
+        
+    def get_availability(self, lid, gid, start, end, user, group=None):            
+        gsr = get_object_or_404(GSR, gid=gid)
+
+        # select a random user from the group if booking wharton gsr
+        if gsr.kind == GSR.KIND_WHARTON and group is not None:
+            wharton_members = group.memberships.filter(is_wharton=True)
+            if (n := wharton_members.count()) == 0:
+                raise APIError("Error: Non Wharton cannot book Wharton GSR")
+            user = wharton_members[randint(0, n-1)].user
+        
+        rooms = self.WBW.get_availability(lid, start, end, user) if gsr.kind == GSR.KIND_WHARTON else self.LBW.get_availability(gid, start, end, user)
+        return {"name": gsr.name, "gid": gsr.gid, "rooms": rooms}
+
+    
+    def get_reservations(self, user, group=None):
+        q = Q(user=user) | Q(reservation__creator=user) if group else Q(user=user)
+        bookings = GSRBooking.objects.filter(q, is_cancelled=False, end__gte=timezone.localtime()).prefetch_related(Prefetch('reservation'))
+
+        if group:
+            ret = []
+            for booking in bookings:
+                data = GSRBookingSerializer(booking).data
+                if booking.reservation.creator == user:
+                    data['room_name'] = f"[Me] {data['room_name']}"
+                else:
+                    data['room_name'] = f"[{group.name}] {data['room_name']}"
+                ret.append(data)
+        else:
+            ret = GSRBookingSerializer(bookings, many=True).data
+        
+        # deal with bookings made directly through wharton (not us)
+        try:
+            wharton_bookings = self.WBW.get_reservations(user) # is this bad? 
```

Comment:
I think this is fine. Let's just keep in mind the `/recent` URL we have in `views.py` which also returns reservations and is used every time a user opens Penn Mobile. Luckily that route doesn't check for reservations made outside of our API so we don't incur this cost during those calls.